### DISCUSSION TITLE
J2 plastic simulation problem

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#16417] [Q&A Modules: Solid mechanics] [answered]
Dear MOOSE experts, 

I have recently been studying the j2 plasticity model in tensor mechanics modulus. Based on the input file [tensor_mechanics_j2plasticity.i](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/j2_plasticity/tensor_mechanics_j2plasticity.i), I have some questions and hope to get your guidance and corrections.

   The following is my understanding and confusion: 
- Kernels: What are the functions implemented?  My understanding is that in this example, the following code means to call [TensorMechanicsAction ](https://mooseframework.inl.gov/syntax/Modules/TensorMechanics/Master/index.html)to use "StressDivergenceTensors" to calculate the balance equation.  I don’t know if I understand right?  Also, does TensorMechanicsAction need to call other kernels to complete this simulation?

`[Kernels]
  [./TensorMechanics]
    displacements = 'x_disp y_disp z_disp'
    use_displaced_mesh = true
  [../]
[]`

For the Materials and AuxKernels modules, my understanding is as follows. If I understand wrongly, I hope to criticize and correct them. 

- Materials: Provide strain, elastic tensor, and plastic yield strength for the kernel. 

- AuxKernels: Calculate physical quantities based on the variables calculated in the kernel for post-processing

Any suggestions or recommendations to fix these problems would be greatly appreciated.
Thanks

Wei Peng

**Comment by WilkAndy** [comment, marked as answer]
Kernels.  Yes the most basic solid mechanics simulation will only use `StressDivergenceTensors`.  However, as your simulation gets more complicated, more Kernels will be needed.  For instance, if your simulation has gravity, then a Kernel (such as `BodyForce`) will be needed.  Or, if your simulation has dynamics, then `DynamicStressDivergenceTensors` will be needed.  Or, if you simulation couples with heat transfer, then Kernels for the heat equation will be needed.

Materials.  Yes, you've got the basic idea.  For solid mechanics simulations, the Materials compute the strain, elasticity tensor, and plastic things, such as plastic strains (for perfect plasticity, plastic yield strength is a real number user input, but for hardening/softening it will be computed by MOOSE based on the plastic strain).

AuxKernels: yes, you've got the general idea correct.  However, for general MOOSE models, these are often used as inputs to other things such as Materials or Kernels, not just for postprocessing.

**Reply by PengWei97** [reply]
Thank you for your quick reply, I got it.

**Reply by Traiwit** [reply]
Hi @WilkAndy,

I'm trying to work on strain-softening, in theory, coh and/or friction angle should be a function of the accumulated plastic strain, is there a way to implement this with moose.

so far I got DP/CM to work, but yet cannot implement softening.
```

[UserObjects]
  [./mc_coh]
    type = TensorMechanicsHardeningConstant
    value = 5e6
  [../]
  [./mc_phi]
    type = TensorMechanicsHardeningConstant
    value = 25
    convert_to_radians = true
  [../]
  [./mc_psi]
    type = TensorMechanicsHardeningConstant
    value = 20
    convert_to_radians = true
  [../]

 ####### Mohr Coulomb
 [./mc]
   type = TensorMechanicsPlasticMohrCoulomb
   cohesion = mc_coh
   friction_angle = mc_phi
   dilation_angle = mc_psi
   yield_function_tolerance = 1E-3
   internal_constraint_tolerance = 1E-4
   mc_tip_smoother = 5e5
   mc_edge_smoother = 20
 [../]
[]

```
Kind regards,
Traiwit

**Reply by WilkAndy** [reply]
Hi @Traiwit - i suggest creating a new Discussion and i'll answer there.
